### DISCUSSION TITLE
pixi built with incorrect Rust version, panics

### DISCUSSION THREAD

**Discussion by mdekstrand** [Homebrew/discussions#5629] [Casks] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.3.23
ORIGIN: https://github.com/Homebrew/brew
HEAD: 42c11a4dedf352229271593de831161071603c66
Last commit: 7 days ago
Core tap JSON: 23 Sep 16:54 UTC
Core cask tap JSON: 23 Sep 16:54 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_AUTO_UPDATE_SECS: 604800
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: nano
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 3.3.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.3.4_1/bin/ruby
CPU: 10-core 64-bit arm_blizzard_avalanche
Clang: 16.0.0 build 1600
Git: 2.46.1 => /opt/homebrew/bin/git
Curl: 8.7.1 => /usr/bin/curl
macOS: 14.7-arm64
CLT: 16.0.0.0.1.1724870825
Xcode: 16.0
Rosetta 2: false
```

### Output of `brew doctor`

```shell
﻿﻿Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some frameworks can be picked up by CMake's build system and will likely
cause the build to fail. To compile CMake, you may wish to move these
out of the way:
  /Library/Frameworks/libcurl.framework
```

### Description of issue

`pixi` is built with the latest version of Rust in Homebrew (`brew info pixi` shows it build-depends on 1.81), which is different from the Rust version specified in `pixi`'s `rust-toolchain` file. This is currently resulting in a panic when solving a large dependency environment:

```
thread 'tokio-runtime-worker' panicked at library/core/src/slice/sort/shared/smallsort.rs:862:5:
user-provided comparison function does not correctly implement a total order
```

I have confirmed that this is a Rust version compatibility issue by rebuilding `pixi` myself using both Rust 1.80.0 (declared in its `rust-toolchain`) and Rust 1.81.0 (equivalent to current Homebrew build) with `rustup` — the 1.81.0 build crashes and the 1.80.0 build does not.

While I believe there is an upstream issue involved here (and I have filed prefix-dev/pixi#2122), building with a version of Rust different from what upstream declares introduces bugs that are not present in the official builds, and thus makes reporting bugs appropriately upstream difficult.

### Relevant casks

pixi

**Comment by p-linnane** [comment]
We only ship versions that are supported by upstream, which for Rust is the latest version (1.81.0). We do not have, nor will we ship, an older unsupported version. This is up to pixi's upstream to implement support for the latest version.

**Reply by mdekstrand** [reply]
Is it feasible to build formulae with rustup (from `rustup-init`) instead of the `rust` formula? The current situation, where `brew install pixi` installs a broken build, does not seem optimal from a user perspective.

**Reply by benglewis** [reply]
@p-linnane I second @mdekstrand 's suggestion. The idea of saying that a particular package declares that it was built with a particular version in it's `rust-toolchain` file and the `homebrew`'s approach is simply to ignore this, try to install it with the latest available `rust` and then allow the end-user to get panics when they are not compatible seems just plain dumb. Either `brew` should compile the package itself with the latest toolchain (which I can't see the benefit from the added complexity) or `brew` should support using `rustup-init` as @mdekstrand suggests to use the correct Rust version together with the precompiled binary. I wouldn't mind trying to open a PR to fix this if there is a chance of it being merged? 🙏

**Reply by SMillerDev** [reply]
> Either brew should compile the package itself with the latest toolchain (which I can't see the benefit from the added complexity)

It already does exactly that.

**Reply by mdekstrand** [reply]
Yep, it isn't a binary compatibility issue — it's a source compatibility issue, and compiling with a newer toolchain than what `pixi` tested with is the problem. To be fair, I think this is likely rare, due to Rust's high degree of backwards compatibility — this specific problem seems to arise from a bug in one of Pixi's dependencies where they implemented `Ord` or `PartialOrd` incorrectly fr some data structure and Rust 1.81 added strictness checks so that sorting with such an incorrect implementation now panics instead of sorting incorrectly, and Pixi reliably hits this bug. Which yes, the root cause is definitely a bug upstream that isn't caught by their current state, but the intermediate state is broken builds.

**Reply by SMillerDev** [reply]
Is there work ongoing upstream to fix this? And could we patch that into the formula?

**Reply by mdekstrand** [reply]
The bug (and reproducers) are being tracked in prefix-dev/pixi#2122, but I haven't chased things through the dependencies to try to see if it's being tracked and fixed there yet.

**Reply by mdekstrand** [reply]
I did a little isolation work and opened conda/rattler#888 as well, things are failing with current upstream `main` branches.

**Comment by benmoss** [comment, marked as answer]
This has been fixed in 0.32.1 in https://github.com/Homebrew/homebrew-core/pull/193199

**Reply by 9764319764319764** [reply]
> This has been fixed in 0.32.1 in https://github.com/Homebrew/homebrew-core/pull/193199

**Comment by 9764319764319764** [comment]
henos

**Reply by 9764319764319764** [reply]
hame
